### PR TITLE
Add DictTensor tensor wrapper subclass

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #146172
* #146101
* #145922
* #141842
* __->__ #141841
* #146052

### CODE REVIEW

File: torch/nested/_internal/cached_tensor.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+
+class CachedTensor(torch.Tensor):
+    @staticmethod
+    def __new__(
+        cls,
+        metadata: Dict[str, torch.Tensor],
+        source_fields: Tuple[str, ...],
+        extra_fields: Tuple[str, ...],
+        target_field: Optional[str] = None,
```
Comment (jbschlosser) [inline torch/nested/_internal/cached_tensor.py:33]: is `target_field` used in NJT? didn't look like it to me but I could be wrong

**Review by jbschlosser — COMMENTED** [review]
Interesting idea to use a tensor subclass here! I think I like it :) I do have some worries about it in its current state though. I find it hard to conceptualize exactly what responsibilities this tensor is handling.

Imo it'd be simpler to reason about to have a tensor subclass that acts like a single tensor but also carries around optional associated metadata. This simplifies the `source_fields`, `target_field`, etc. distinction: there'd be a single tensor this instance acts as and a set of other tensors carted around.

It's more complex and odd to reason about to have the possibility of multiple `source_field`s or tensors that are essentially equivalent and interchangeable (?) I'm not really sure how this is utilized in practice for NJT but it doesn't seem like it could be seamless anyway (correct me if I'm wrong), so might as well remove the complexity.

TL;DR: the responsibility of considering two tensors as equivalent (and the complexity / concepts needed) feels outside of the scope of a cached tensor object and makes it harder to reason about / grok. If it's possible to boot this to a layer above in NJT, I'd prefer that

File: torch/nested/_internal/cached_tensor.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+#
+# Registering this decorator for an op allows CachedTensor's torch dispatch to
+# redirect calls to that op to your own implementations.
+# For NestedTensor we rely on this behavior for factory functions.
+def register_cached_tensor_func(aten_op: Any) -> Callable:
+    def wrapper(func: Callable) -> Callable:
+        _func_registry[aten_op] = func
+        return func
+
+    return wrapper
```
Comment (jbschlosser) [inline torch/nested/_internal/cached_tensor.py:152]: I think it's an interesting way to customize the behavior of this class based on what the parent subclass needs. For the lazy: what in particular is this utilized for for NJT / torch.compile?

File: torch/nested/_internal/cached_tensor.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        else:
+            raise AttributeError(
+                f"{type(self).__name__} object has no attribute '{name}'"
+            )
+
+    def __tensor_flatten__(self) -> Tuple[List[str], Dict[str, Any]]:
+        ctx = {
+            "source_fields": self.source_fields,
+            "extra_fields": self.extra_fields,
+        }
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:85]: Whoops

File: torch/nested/_internal/cached_tensor.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        target_field: Optional[str] = None,
+    ):
+        self.source_fields = source_fields
+        self.extra_fields = extra_fields
+        self.all_fields = source_fields + extra_fields
+        self.target_field = target_field
+        self.metadata = metadata
+
+        # Compile only
+        self.nested_int_ref: Any = None
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:65]: I guess moving it to weaktensorkeydict on FakeMode from CachedTensor to nested int ref would fully clean up this class

File: torch/nested/_internal/cached_tensor.py (hunk 2 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:152]: Specifically, NJT has two constructors from_padded and from_jagged that rely on this. Both require CachedTensor as input.

File: torch/nested/_internal/cached_tensor.py (hunk 1 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:33]: The original use case here was if a NJT's metadata holds both the lengths and offsets. If the user calls .offsets(), we'd want to return a metadata with target_field as offsets, and analgously for lengths.

I ended up updating .offsets()/._offsets to just return plain tensors, so this logic is no longer used by NJT today, but I imagine it would be if someone wanted a .offsets_with_cache() API.

File: test/test_nestedtensor.py [hunk 5]
```diff
@@ -8654,6 +8654,111 @@ def test_with_factor(self):
         self.assertTrue(a * 2 == 2 * a)
 
 
+from torch.nested._internal.cached_tensor import CachedTensor
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nestedtensor.py:8657]: I would agree with you haha, sadly it's hard to behave like a true CachedTensor without metadata mutation, so  suggestions welcome/happy to bikeshed some more.

Also given the approach of handling the source + extra fields distinction at a different layer, it could even be something like `ContainerTensor`.

**Comment by soulitzer** [comment]
> Interesting idea to use a tensor subclass here! I think I like it :) I do have some worries about it in its current state though. I find it hard to conceptualize exactly what responsibilities this tensor is handling.

> Imo it'd be simpler to reason about to have a tensor subclass that acts like a single tensor but also carries around optional associated metadata. This simplifies the source_fields, target_field, etc. distinction: there'd be a single tensor this instance acts as and a set of other tensors carted around.

> It's more complex and odd to reason about to have the possibility of multiple source_fields or tensors that are essentially equivalent and interchangeable (?) I'm not really sure how this is utilized in practice for NJT but it doesn't seem like it could be seamless anyway (correct me if I'm wrong), so might as well remove the complexity.

> TL;DR: the responsibility of considering two tensors as equivalent (and the complexity / concepts needed) feels outside of the scope of a cached tensor object and makes it harder to reason about / grok. If it's possible to boot this to a layer above in NJT, I'd prefer that

This is great feedback, let me play around with some more :)

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -8654,6 +8654,111 @@ def test_with_factor(self):
...
+                self.assertEqual(
+                    cloned_cached_tensor.metadata[key], cached_tensor.metadata[key]
+                )
+                self.assertFalse(
+                    cloned_cached_tensor.metadata[key] is cached_tensor.metadata[key]
+                )
+
+        # After leaving the context, clone behaves as it did before.
+        cloned_cached_tensor = cached_tensor.clone()
+        self.assertFalse(isinstance(cloned_cached_tensor, CachedTensor))
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nestedtensor.py:8732]: Another example actually where clone does not produce an object with the same type is `nn.Parameter`

```
>>> a = torch.nn.Parameter(torch.tensor(1.))
>>> isinstance(a.clone(), torch.nn.Parameter)
False
```

**Comment by soulitzer** [comment]
Ready for review again!

Updates: 
- no longer handling the source/extra distinction at this layer (those are handled at NT level instead)
- instead of holding weakref on the CachedTensor itself, FakeTensorMode holds a WeakTensorKeyDictionary

File: torch/nested/_internal/cached_tensor.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    #
+    # When used this way (1) it is a convenient way to keep around metadata
+    # related to a tensor, without having to laboriously thread those extra metadata
+    # around, e.g. through aten signatures. (2) allows one to trigger custom
+    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
+    # otherwise because the subclass constructor op itself usually does not take the
+    # subclass itself as input!). See NestedTensor for an example.
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
```
Comment (jbschlosser) [inline torch/nested/_internal/cached_tensor.py:30]: why does this need to be here?

File: torch/nested/_internal/cached_tensor.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    def __new__(
+        cls,
+        metadata: Dict[str, Optional[torch.Tensor]],
+        source_field: str,
+    ) -> "CachedTensor":
```
Comment (jbschlosser) [inline torch/nested/_internal/cached_tensor.py:35]: Okay I'm loving how much simpler this class is now! There's still one thing bothering me (from the perspective of reviewing this in isolation without looking much in detail at how it's used for NJT): the concept of a specifying a dict and source field that refers to an object in that dict seems overly restrictive and a bit awkward.

Ideally I'd want to see something like:
```
CacheTensor(t: torch.Tensor, cache: Dict[str, Optional[torch.Tensor]])
```

where the `CacheTensor` operates like `t` but also contains the extra `cache` stuff alongside it, and notably `cache` doesn't contain `t`. This seems about as clean as possible.

Would this cause problems with the way NJT wants to think about `source_field`? It feels like this should still accomplish what NJT needs, albeit slightly differently, but lmk if you think there are unsolvable perf issues with this.

File: torch/nested/_internal/cached_tensor.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        kwargs["requires_grad"] = source.requires_grad  # type: ignore[assignment]
+        kwargs["dtype"] = source.dtype  # type: ignore[assignment]
+        out = torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+        out.metadata = metadata
+        out.source_field = source_field
+        return out
+
+    def __repr__(self) -> str:  # type: ignore[override]
+        return f"CachedTensor({repr(self.metadata[self.source_field])})"
```
Comment (jbschlosser) [inline torch/nested/_internal/cached_tensor.py:53]: I'll suggest `CacheTensor` as a slightly tweaked name since it's a tensor that contains additional cache info.

**Review by jbschlosser — COMMENTED** [review]
Nice work! I think this most recent iteration is much simpler and closer to be being approvable. Just got one more thing bugging me (open to being convinced otherwise) and I think the name should be tweaked a bit and then it's good to go imo.

File: test/test_nestedtensor.py [hunk 10]
```diff
@@ -8654,6 +8654,111 @@ def test_with_factor(self):
         self.assertTrue(a * 2 == 2 * a)
 
 
+from torch.nested._internal.cached_tensor import CachedTensor
+
+
+class TestCachedTensor(torch.testing._internal.common_utils.TestCase):
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_nestedtensor.py:8912]: bumping this - should probably still move

File: torch/nested/_internal/cached_tensor.py (hunk 7 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:30]: From talking with @bdhirsh its standard practice to prevent dynamo from randomly tracing this frame and erroring out in a weird way. (It's also used in DTensor.)

Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:30]: This reminds me to add a test for constructing CachedTensor in the middle of the graph (which we may rely on for NestedTensor)

File: torch/nested/_internal/cached_tensor.py (hunk 9 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:53]: Sounds good!

File: torch/nested/_internal/cached_tensor.py (hunk 8 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:35]: The main difference I see between the two approaches are that one requires the source field to have a name, the other does not. 

For NJT requiring the source field to have a name is useful because it allows me to access every field on the CacheTensor including the source field in a uniform way.

File: torch/nested/_internal/cached_tensor.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                f"{type(self).__name__} object has no attribute '{name}'"
+            )
+        return self.metadata[name]
+
+    def __tensor_flatten__(self) -> Tuple[List[str], Dict[str, Any]]:
+        ctx = {
+            "source_field": self.source_field,
+            "all_fields": list(self.metadata.keys()),
+        }
+        return [k for k, v in self.metadata.items() if v is not None], ctx
```
Comment (jbschlosser) [inline torch/nested/_internal/cached_tensor.py:67]: note that the order is non-deterministic here due to the use of `items()`. Is this a problem? @bdhirsh

File: torch/nested/_internal/cached_tensor.py (hunk 8 above)
Comment (jbschlosser) [reply torch/nested/_internal/cached_tensor.py:35]: Hm I'm not super convinced, especially because I want to consider this class outside of an NJT context. The source field thing makes CacheTensor usage marginally simpler within NJT but more awkward for other use cases imo. Treating the source field and some other cached objects uniformly seems pretty specific to the NJT use case, and arguably should be handled at the NJT layer.

> For NJT requiring the source field to have a name is useful because it allows me to access every field on the CacheTensor including the source field in a uniform way.

You can maintain a name within NJT, no? e.g. store canonical source field for a given NJT instance:
```
class NJT:
    ...
    source_field = "_device_offsets"  # for example
    source = CacheTensor(source_tensor, other_tensors)

    # custom getattr returns CacheTensor's t for source_field and indexes other_tensors for other fields.
    # it's less uniform from the perspective of NJT usage but places responsibilities at the correct layer imo
```

From what I can tell, this isn't too hard to deal with in NJT but it does make CacheTensor even dumber and straightforward (in a good way imo)

File: torch/nested/_internal/cached_tensor.py (hunk 11 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:67]: Is this still the case in Python versions >=  3.7?

https://stackoverflow.com/questions/14956313/why-is-dictionary-ordering-non-deterministic

File: torch/nested/_internal/cached_tensor.py (hunk 8 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:35]: Is the concern is that CacheTensor in its current state isn't constructed like one would expect a tensor named CacheTensor? e.g. in this model one would expect CacheTensor to be first and foremost a tensor, but has additional fields stored in a dict. Given this, I think the concern about its construction makes sense.

Could I alternatively propose that we changed the name of the tensor to `SimpleTensorDict`? In that model this tensor is more like a wrapper over a dict, in which every field is seen uniformly. In this world, the current way of construction should be most natural.

Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:35]: Thinking out side of the NJT context, perhaps a reason to favor `SimpleTensorDict` over `CacheTensor` is that
CacheTensor doesn't actually benefit much from being first and foremost a tensor, because it immediately unwraps upon any operators call.  A container-like name of  simple-tensor-dict makes it clearer that this is more of a container object than a tensor you are meant to operate on.

Comment (jbschlosser) [reply torch/nested/_internal/cached_tensor.py:35]: > Is the concern is that CacheTensor in its current state isn't constructed like one would expect a tensor named CacheTensor? e.g. in this model one would expect CacheTensor to be first and foremost a tensor, but has additional fields stored in a dict. Given this, I think the concern about its construction makes sense.

Yep you nailed it here. Trying to have the impl match the mental model of such a class (for reasoning / maintenance purposes mainly - much easier to verify things with clear semantics).

> Could I alternatively propose that we changed the name of the tensor to SimpleTensorDict? In that model this tensor is more like a wrapper over a dict, in which every field is seen uniformly. In this world, the current way of construction should be most natural.
>
> Thinking out side of the NJT context, perhaps a reason to favor SimpleTensorDict over CacheTensor is that
CacheTensor doesn't actually benefit much from being first and foremost a tensor, because it immediately unwraps upon any operators call. A container-like name of simple-tensor-dict makes it clearer that this is more of a container object than a tensor you are meant to operate on.

Okay I generally like where this is going- if you really want to act on a dictionary cache, use an object that fits that model. How would the torch_dispatch for `SimpleTensorDict` work - would it still unwrap for a specific key?

File: torch/nested/_internal/cached_tensor.py (hunk 11 above)
Comment (jbschlosser) [reply torch/nested/_internal/cached_tensor.py:67]: hm guess not :) so we're good as long as we don't support older python versions than that

File: torch/nested/_internal/cached_tensor.py (hunk 8 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:35]: That is a good point - It is true that the current behavior is not symmetric wrt all the field... maybe what I really want to happen is to ban all operations.

Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:35]: > maybe what I really want to happen is to ban all operations.

This should be fine except for the purpose of compile, I need to support 'detach'. WDYT?

Comment (jbschlosser) [reply torch/nested/_internal/cached_tensor.py:35]: > This should be fine except for the purpose of compile, I need to support 'detach'. WDYT?

Sure, so it's only purpose is to store tensors into a dict in a way that matches the "traceable wrapper subclass" interface? Does `detach()` detach all the tensors in the dict?

Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:35]: > Sure, so it's only purpose is to store tensors into a dict in a way that matches the "traceable wrapper subclass" interface?

Exactly

>  Does detach() detach all the tensors in the dict?

I think we want to return a fresh NestedTensor object, but I'm not completely sure if I want to detach everything inside.

What tips me slightly toward not detaching overall is because generally for NT's size comparison logic, we'd prefer if the instance of the offsets, etc is the same. (I'm not sure it actually errors either when I do detach, but didn't  investigate further)

For other things it seems like it shouldn't matter:

- From the perspective of compile it seems okay not to detach because we already have the precedence of NJT not detaching offsets when the outer NJT is detached.

- From the perspective of autograd it shouldn't matter either here since detach is a no-op underneath autograd.

One might argue that one would expect a dict-container-like tensor replicate the outer commands into all the inner commands, which I don't disagree with, but at the same time, I don't really consider detach() truly public API either 
because I don't expect the user to be randomly calling into detach when no other op is supported.

File: torch/nested/_internal/cached_tensor.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    ) -> torch.Tensor:
+        # Ensure that any registered ops are loaded
+        import torch.nested._internal.ops  # noqa: F401
+
+        if kwargs is None:
+            kwargs = {}
+
+        if op is torch.ops.aten.detach.default:
+            # detach is needed for torch.compile
+            return args[0]
```
Comment (soulitzer) [inline torch/nested/_internal/cached_tensor.py:85]: Reminder to rewrap to fresh tensor.

File: torch/nested/_internal/cached_tensor.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        out = torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+        out.metadata = metadata
+        out.source_field = source_field
+        return out
+
+    def __repr__(self) -> str:  # type: ignore[override]
+        return f"CachedTensor({repr(self.metadata[self.source_field])})"
+
+    def __getattr__(self, name: str) -> Optional[torch.Tensor]:
```
Comment (bdhirsh) [inline torch/nested/_internal/cached_tensor.py:52]: Is this custom getattr mainly to make sure that `cache_tensor.key` secretly maps to `cache_tensor.metadata.key`, so the source information of inner tensors is accurate?

That makes sense, although I'm not sure how good dynamo's handling of custom getattr's is. Do you ever expect attribute accesses on a `CachedTensor` to happen from user python code, that dynamo needs to inline?

File: torch/nested/_internal/cached_tensor.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+            "source_field": self.source_field,
+            "all_fields": list(self.metadata.keys()),
+        }
+        return [k for k, v in self.metadata.items() if v is not None], ctx
+
+    @staticmethod
+    def __tensor_unflatten__(
+        inner_tensors: Dict, meta: Dict, outer_size: Any, outer_stride: Any
+    ) -> "CachedTensor":
+        inner_tensors = {k: inner_tensors.get(k) for k in meta["all_fields"]}
```
Comment (bdhirsh) [inline torch/nested/_internal/cached_tensor.py:70]: aren't `inner_tensors.keys()` and `meta["all_fields"]` identical / redundant? Or is it possible for them to be different?

File: torch/nested/_internal/cached_tensor.py (hunk 13 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:45]: > Is this custom getattr mainly to make sure that cache_tensor.key secretly maps to cache_tensor.metadata.key, so the source information of inner tensors is accurate?

Yeah, this was mainly to get `__tensor_flatten__` to work

> That makes sense, although I'm not sure how good dynamo's handling of custom getattr's is. Do you ever expect attribute accesses on a CachedTensor to happen from user python code, that dynamo needs to inline?

Nah, today I expect CachedTensor to be purely private.

File: torch/nested/_internal/cached_tensor.py (hunk 14 above)
Comment (soulitzer) [reply torch/nested/_internal/cached_tensor.py:62]: Inner tensors keys are incomplete because they only contain the k's that are non-None. (I'm not sure if we can have inner tensors that are None.)

**Comment by soulitzer** [comment]
Updates:
- Renamed from CachedTensor to DictTensor (see this long discussion for context https://github.com/pytorch/pytorch/pull/141841#discussion_r1887678558)
- Moved the test from test_nestedtensor to a separate file (test_dict_tensor)
- The shape of the tensor is (0,) to avoid trying to copy the metadata of some arbitrary inner tensor.

File: test/test_nestedtensor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_nestedtensor.py]: Done!

File: torch/nested/_internal/dict_tensor.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
```
Comment (Skylion007) [inline torch/nested/_internal/dict_tensor.py:2]: I'm surprised we allow star imports.

File: torch/nested/_internal/dict_tensor.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        _func_registry = old_registry
+
+
+# Note: [ CacheTensor open registry ]
+#
+# Registering this decorator for an op allows DictTensor's torch dispatch to
+# redirect calls to that op to your own implementations.
+# For NestedTensor we rely on this behavior for factory functions.
+def register_dict_tensor_func(aten_op: Any) -> Callable:
+    def wrapper(func: Callable) -> Callable:
```
Comment (Skylion007) [inline torch/nested/_internal/dict_tensor.py:109]: This erases the return type and params of the callable, you either want to use ParamSpec or a TypeVar
